File: nim/loops.nim
import os
import random
import strutils
import times
import math

# variables
let start = epochTime()
let iterations = parseInt(paramStr(1))
const sampleSize = 23
var count = 0

proc roundTo(number: float, places: float): float =
  let shifter = pow(10, places)
  int(number * shifter) / int(shifter)

# setup
randomize()
for i in countup(1, iterations):
  type Sample = array[1..365, bool]
  var sample: Sample
  for s in countup(1, sampleSize):
    let day = random(1..365)
    if sample[day]:
      inc(count)
      break
    else:
      sample[day] = true

# calcs
let percent = count / iterations * 100
let fin = epochTime()
let diff = fin - start

# output
echo "iterations: " & $iterations
echo "sample-size: " & $sampleSize
echo "percent: " & $roundTo(percent, 2)
echo "seconds: " & $roundTo(diff, 3)

File: solutions/kotlin-native/loops.kt
// import java.util.Random
import kotlin.random.*
import kotlin.system.*
import kotlin.math.*

fun main(args: Array<String>) {
  val iterations = args[0].toInt()
  simulate(iterations)
}

fun simulate(iterations: Int) {
  // val start = System.currentTimeMillis()
  val start = getTimeMillis()
  val sample_size = 23

  var count = 0
  for (i in 1..iterations) {
    val data = IntArray(365)
    for (n in 1..sample_size) {
      val number = Random.nextInt(0,365)
      if (data[number] == 1) {
        count++
        break
      } else {
        data[number] = 1
      }
    }
  }
  println("iterations: ${iterations}")
  println("sample-size: ${sample_size}")
  val percent = count * 100.0 / iterations
  val formattedPercent = round(percent * 100) / 100
  println("percent: ${formattedPercent}")
  val end = getTimeMillis()
  val diff = (end - start) / 1000.0
  println("seconds: ${diff}")
}
